File: grooves-api/src/main/groovy/com/github/rahulsom/grooves/api/events/JoinEvent.java
package com.github.rahulsom.grooves.api.events;

import com.github.rahulsom.grooves.api.AggregateType;

/**
 * Creates a join from {@link AggregateT} to {@link JoinedAggregateT} that had not existed earlier.
 *
 * @param <AggregateT>       The Aggregate that has been linked to the {@link JoinedAggregateT}
 * @param <EventIdT>         The Type for Event's {@link #getId} field
 * @param <EventT>           The parent event type
 * @param <JoinedAggregateT> The target aggregate
 *
 * @author Rahul Somasunderam
 */
public interface JoinEvent<
        AggregateT extends AggregateType,
        EventIdT,
        EventT,
        JoinedAggregateT extends AggregateType>
        extends BaseEvent<AggregateT, EventIdT, EventT> {

    JoinedAggregateT getJoinAggregate();

    void setJoinAggregate(JoinedAggregateT rollupAggregate);
}

File: grooves-types/src/main/kotlin/com/github/rahulsom/grooves/api/events/JoinEvent.kt
package com.github.rahulsom.grooves.api.events

import org.reactivestreams.Publisher

/**
 * Creates a join from [AggregateT] to [JoinedAggregateT] that had not existed earlier.
 *
 * @param [AggregateT] Aggregate this event applies to
 * @param [EventIdT] The type for [BaseEvent.id]
 * @param [EventT] The base class for all events in the hierarchy for [AggregateT].
 * @param [JoinedAggregateT] The target aggregate
 *
 * @author Rahul Somasunderam
 */
interface JoinEvent<
        AggregateT,
        EventIdT,
        EventT,
        JoinedAggregateT> :
    BaseEvent<AggregateT, EventIdT, EventT> {

    /**
     * An observable that points to the aggregate to which a join is being performed.
     *
     * This join is unidirectional, but a good system will always create a pair of events to make all joins bidirectional.
     * That makes it possible for a system at a later date to change it's problem space and answer questions that unidirectional joins can't.
     * Your directionality should match your plan in [DisjoinEvent]
     */
    val joinAggregateObservable: Publisher<JoinedAggregateT>
}